Implemented APIv3 endpoints & authentication

aurelienhess opened the pull request:
Added a client for SaaS API

aurelienhess committed:
Implemented APIv3 endpoints & authentication

Added a client for SaaS API

aurelienhess committed:
Merge branch 'master' into patch-33

aurelienhess committed:
Inherited from 2.4 and added docstring

HoneyBaked commented on modules/vectra.py:
```diff
@@ -63,6 +65,29 @@ def api_validator(self, *args, **kwargs):
+            # Once the token is refreshed, we can retry the operation.
+            return func(self, *args, **kwargs)
+        except HTTPTooManyRequestsException:
+            sleep(1)
```
Note, this will retry forever, if the client code continues to get the caught exception.  This is likely ok for now.

HoneyBaked commented on modules/vectra.py:
```diff
@@ -2670,3 +2696,283 @@ def delete_group(self, group_id=None):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        # Get the OAuth2 token
+        r_dict = self._get_oauth_token(url, client_id, client_secret).json()
```
We should avoid having an external request, such as an http request, in the init of the class.  init should likely initialize the access_token as None.

1. init sets self.access_token = None, self.refresh_token
2. _request is called, checks if self.access_token is valid, if not set invoke _get_oauth_token and set self.access_token, self.refresh_token
3. we likely need to update _request to handle authentication exceptions.  If _request fails due to auth issues, then refresh the token...

HoneyBaked commented on modules/vectra.py:
```diff
@@ -63,6 +65,29 @@ def api_validator(self, *args, **kwargs):
+    def wrapper(self, *args, **kwargs):
+        try:
+            return func(self, *args, **kwargs)
+        except HTTPUnauthorizedException:
```
This exception should eventually give up.  It's possible refresh_oauth_token may fail.  Do we need to check the _refresh_oauth_token's result here, and raise (fail) if we are unable to refresh the token?

aurelienhess committed:
Moves OAuth out of init function

- Moved the initial OAuth token gathering out of the init function
- Added time-based validity checks so that we don't retry indefinitely when getting 401 errors
- Moved all functions that are only needed in the scope of a class within that class
- Minor corrections

HoneyBaked commented on modules/vectra.py:
```diff
@@ -54,22 +56,35 @@ def request_handler(self, *args, **kwargs):
-
+        except HTTPUnauthorizedException as e:
+            # To calculate token expiration we take 10s margin
+            if not self.refresh_token or (self.access_token_validity-time() > 10.0 and self.refresh_token_validity-time()>10.0):
```
I would check your formatting throughout this file.  Running black should help.  We have cases where we have the operator missing spaces to the right and/or left.

It may be good to run black against this file as a separate PR:
https://pypi.org/project/black/
